examples/self-modifying-code/context.ts

```diff
@@ -0,0 +1,101 @@
+import fs from "fs/promises";
+import path from "path";
+
+interface FileSystemError extends Error {
+  code?: string;
+}
+
+export interface AgentContext {
+  goalState: string;
+  history: {
+    timestamp: Date;
+    action: string;
+    result: "success" | "failure";
+    details: string;
+  }[];
+}
+
+function serializeContext(context: AgentContext): string {
+  return JSON.stringify(
+    context,
+    (_key, value) => {
+      if (value instanceof Date) {
+        return value.toISOString();
+      }
+      return value;
+    },
+    2,
+  );
+}
+
+function deserializeContext(json: string): AgentContext {
+  return JSON.parse(json, (key, value) => {
+    if (key === "timestamp" && typeof value === "string") {
+      return new Date(value);
+    }
+    return value;
+  });
+}
+
+export async function readContext(contextFile: string): Promise<AgentContext> {
```

Comment:
You might be able to simplify all this to just be `import context from 'context.json;`. Since it modifies the source, when the new process starts up it will get the updated version of the context.